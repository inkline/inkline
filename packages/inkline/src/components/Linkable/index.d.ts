import { Component } from 'vue';

declare const ILinkable: Component;

export default ILinkable;
