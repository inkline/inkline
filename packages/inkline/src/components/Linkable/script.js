import {
    ActivePropertyMixin,
    DisabledPropertyMixin,
    AttributesProviderMixin,
    ClassesProviderMixin,
    TabIndexPropertyMixin,
} from '@inkline/inkline/src/mixins';

export default {
    name: 'ILinkable',
    mixins: [
        ActivePropertyMixin,
        TabIndexPropertyMixin,
        DisabledPropertyMixin,

        AttributesProviderMixin,
        ClassesProviderMixin
    ],
    data() {
        return {
            routerComponent: this.$nuxt ? 'nuxt-link' : 'router-link'
        };
    },
    props: {
        tag: {
            type: String,
            default: 'a'
        },
        tabindex: {
            type: [Number, String],
            default: -1
        }
    },
    computed: {
        isTag() {
            return this.attributes.to ? this.routerComponent : this.attributes.href ? 'a' : this.tag;
        },
        isComponent() {
            return this.isTag === this.routerComponent;
        }
    },
    methods: {
        onClick(event) {
            this.$emit('click', event);
        }
    }
};
