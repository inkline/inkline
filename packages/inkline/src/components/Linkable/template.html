<component v-if="isComponent" :is="isTag" class="item" :class="classes" v-bind="attributes" :disabled="isDisabled" :aria-disabled="isDisabled"
           @click.native="onClick">
    <slot></slot>
</component>
<component v-else :is="isTag" class="item" :class="classes" v-bind="attributes" :disabled="isDisabled" :aria-disabled="isDisabled"
           @click="onClick">
    <slot></slot>
</component>
