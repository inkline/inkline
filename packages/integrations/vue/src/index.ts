export * from './composables';
export * from './constants';
export * from './plugins';
export * from './plugin';
export * from './types';
