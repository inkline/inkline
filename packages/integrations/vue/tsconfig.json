{
    "extends": "../../tooling/typescript-config/tsconfig.json",
    "compilerOptions": {
        "types": [
            "vite/client",
            "vitest/globals"
        ]
    },
    "include": [
        "src/**/*.ts",
        "tsup.config.ts",
        "vitest.config.ts"
    ]
}
