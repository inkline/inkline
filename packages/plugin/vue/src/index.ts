export * from './plugin';
export * from './symbols';
