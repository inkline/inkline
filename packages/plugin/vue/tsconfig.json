{
    "extends": "../../tooling/typescript-config/tsconfig.json",
    "compilerOptions": {
        "paths": {
            "@inkline/types": ["../../types/src/*"]
        },
        "types": [
            "vite/client",
            "vitest/globals"
        ]
    },
    "include": [
        "src/**/*.ts",
        "tsup.config.ts",
        "vite.config.ts"
    ]
}
