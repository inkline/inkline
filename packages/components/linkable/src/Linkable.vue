<script lang="ts">
import { Component, PropType, resolveComponent } from 'vue';
import { computed, defineComponent } from 'vue';
import { useOptions } from '@inkline/composables';

const componentName = 'Linkable';

export default defineComponent({
    name: componentName,
    props: {
        /**
         * Renders the component as an anchor link with a `href` attribute
         * @type String
         * @default undefined
         * @name to
         */
        href: {
            type: String,
            default: undefined
        },
        /**
         * Set the HTML tag to be used for rendering the linkable
         * @type String
         * @default button
         * @name tag
         */
        tag: {
            type: String,
            default: 'a'
        },
        /**
         * Renders the component as a Router Link component with a `to` attribute
         * @type String
         * @default undefined
         * @name to
         */
        to: {
            type: [String, Object] as PropType<string | object>,
            default: undefined
        }
    },
    setup(props) {
        const { options } = useOptions();

        const isRouterLink = computed(() => {
            return props.to;
        });

        const isLink = computed(() => {
            return props.href;
        });

        const renderTag = computed(() => {
            if (props.href) {
                return 'a';
            } else {
                return props.tag;
            }
        });

        const renderComponent = computed(() => {
            const routerComponent = options.value.router?.component;
            if (props.to && routerComponent) {
                if (typeof routerComponent === 'string') {
                    return resolveComponent(routerComponent) as Component;
                } else {
                    return routerComponent;
                }
            }
        });

        return {
            isRouterLink,
            isLink,
            renderComponent,
            renderTag
        };
    }
});
</script>

<template>
    <Component
        v-if="isRouterLink"
        v-bind="$attrs"
        :is="renderComponent"
        :tag="renderTag"
        :to="to"
        role="link"
    >
        <slot />
    </Component>
    <a v-else-if="isLink" v-bind="$attrs" :href="href" role="link">
        <slot />
    </a>
    <button v-else-if="renderTag === 'button'" v-bind="$attrs" role="link">
        <slot />
    </button>
    <input v-else-if="renderTag === 'input'" v-bind="$attrs" type="button" role="link" />
    <Component v-else v-bind="$attrs" :is="renderTag" role="link">
        <slot />
    </Component>
</template>
