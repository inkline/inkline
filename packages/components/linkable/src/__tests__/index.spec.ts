import { render } from '@testing-library/vue';
import Linkable from '../Linkable.vue';
import { createTestingInklineOptionsProvide } from '@inkline/test-utils';

describe('Components', () => {
    describe('Linkable', () => {
        const stubs = {
            RouterLink: true
        };

        const props = {
            href: 'https://example.com',
            tag: 'div',
            to: '/home'
        };

        it('should be named correctly', () => {
            expect(Linkable.name).toEqual('Linkable');
        });

        it('should render correctly', () => {
            const wrapper = render(Linkable, {
                props,
                global: {
                    stubs,
                    provide: {
                        ...createTestingInklineOptionsProvide()
                    }
                }
            });
            expect(wrapper.html()).toMatchSnapshot();
        });

        it('should render as a router link when "to" prop is provided', () => {
            const wrapper = render(Linkable, {
                props: {
                    to: '/home'
                },
                global: {
                    stubs,
                    provide: {
                        ...createTestingInklineOptionsProvide()
                    }
                }
            });
            expect(wrapper.container.firstChild).toHaveAttribute('role', 'link');
        });

        it('should render as an anchor link when "href" prop is provided', () => {
            const wrapper = render(Linkable, {
                props: {
                    href: 'https://example.com'
                },
                global: {
                    stubs,
                    provide: {
                        ...createTestingInklineOptionsProvide()
                    }
                }
            });
            expect(wrapper.container.firstChild).toHaveAttribute('href', 'https://example.com');
        });

        it('should render as a button when "tag" prop is "button"', () => {
            const wrapper = render(Linkable, {
                props: {
                    tag: 'button'
                },
                global: {
                    stubs,
                    provide: {
                        ...createTestingInklineOptionsProvide()
                    }
                }
            });
            expect(wrapper.container.firstChild).toHaveAttribute('role', 'link');
        });

        it('should render as a custom tag when "tag" prop is provided', () => {
            const wrapper = render(Linkable, {
                props: {
                    tag: 'div'
                },
                global: {
                    stubs,
                    provide: {
                        ...createTestingInklineOptionsProvide()
                    }
                }
            });
            expect(wrapper.container.firstChild!.nodeName).toBe('DIV');
        });
    });
});
