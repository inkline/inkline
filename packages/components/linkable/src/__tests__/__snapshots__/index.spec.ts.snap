// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components > Linkable > should render correctly 1`] = `"<router-link-stub tag="a" to="/home" role="link"></router-link-stub>"`;
